pkg/generate/app/pipeline.go

```diff
@@ -15,34 +16,69 @@ import (
 	deploy "github.com/openshift/origin/pkg/deploy/api"
 	image "github.com/openshift/origin/pkg/image/api"
 	route "github.com/openshift/origin/pkg/route/api"
+	"github.com/openshift/origin/pkg/util/docker/dockerfile"
 )
 
-// Pipeline holds components.
-type Pipeline struct {
-	From string
+// A PipelineBuilder creates Pipeline instances.
+type PipelineBuilder interface {
+	NewBuildPipeline(string, *ComponentMatch, *SourceRepository) (*Pipeline, error)
+	NewImagePipeline(string, *ComponentMatch) (*Pipeline, error)
+}
 
-	InputImage *ImageRef
-	Build      *BuildRef
-	Image      *ImageRef
-	Deployment *DeploymentConfigRef
-	Labels     map[string]string
+// NewPipelineBuilder returns a PipelineBuilder using name as a base name. A
+// PipelineBuilder always creates pipelines with unique names, so that the
+// actual name of a pipeline (Pipeline.Name) might differ from the base name.
+// The pipelines created with a PipelineBuilder will have access to the given
+// environment. The boolean outputDocker controls whether builds will output to
+// an image stream tag or docker image reference.
+func NewPipelineBuilder(name string, environment Environment, outputDocker bool) PipelineBuilder {
+	return &pipelineBuilder{NewUniqueNameGenerator(name), environment, outputDocker}
 }
 
-// NewImagePipeline creates a new pipeline with components that are not expected
-// to be built.
-func NewImagePipeline(from string, image *ImageRef) (*Pipeline, error) {
-	return &Pipeline{
-		From:  from,
-		Image: image,
-	}, nil
+type pipelineBuilder struct {
+	nameGenerator UniqueNameGenerator
+	environment   Environment
+	outputDocker  bool
 }
 
 // NewBuildPipeline creates a new pipeline with components that are expected to
 // be built.
-func NewBuildPipeline(from string, input *ImageRef, outputDocker bool, strategy *BuildStrategyRef, env Environment, source *SourceRef) (*Pipeline, error) {
-	name, ok := NameSuggestions{source, input}.SuggestName()
-	if !ok {
-		return nil, ErrNameRequired
+func (pb *pipelineBuilder) NewBuildPipeline(from string, resolvedMatch *ComponentMatch, sourceRepository *SourceRepository) (*Pipeline, error) {
+	input, err := InputImageFromMatch(resolvedMatch)
+	if err != nil {
+		return nil, fmt.Errorf("can't build %q: %v", from, err)
+	}
+	if !input.AsImageStream {
+		msg := "Could not find an image stream match for %q. Make sure that a Docker image with that tag is available on the node for the build to succeed."
+		glog.Warningf(msg, resolvedMatch.Value)
```

Comment:
I would vote for changing that in a separate PR. I'm keeping external behavior unchanged with this refactor PR.